hostplumber/controllers/hostconfig_controller.go

```diff
@@ -0,0 +1,486 @@
+	NodeName    string
+	currentSpec *plumberv1.HostConfigSpec
+}
+
+// +kubebuilder:rbac:groups=plumber.k8s.pf9.io,resources=hostconfigs,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=plumber.k8s.pf9.io,resources=hostconfigs/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=*,resources=*,verbs=*
+
+func (r *HostConfigReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+	log := r.Log.WithValues("hostconfig", req.NamespacedName)
+
+	var specApplied bool = false
+	var hostConfigReq = plumberv1.HostConfig{}
+	if err := r.Get(ctx, req.NamespacedName, &hostConfigReq); err != nil {
+		log.Error(err, "unable to fetch HostConfig")
+		fmt.Println("Unable to fetch new HostConfig CRD!!!")
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	r.currentSpec = &hostConfigReq.Spec
+
+	sriovConfigList := r.currentSpec.SriovConfig
+
+	for _, sriovConfig := range sriovConfigList {
+		var pfName string
+		var pfList []string
+		var err error
+		if sriovConfig.PfName != nil {
+			fmt.Printf("Configuring via PF: %s\n", *sriovConfig.PfName)
+			pfName = *sriovConfig.PfName
+			pfList = append(pfList, pfName)
+		} else if sriovConfig.PciAddr != nil {
+			fmt.Printf("Now configuring via PCI: %s\n", *sriovConfig.PciAddr)
+			pfName, err = getPfNameForPciAddr(*sriovConfig.PciAddr)
+			if err != nil {
+				fmt.Printf("Failed to find PciAddr %s\n", *sriovConfig.PciAddr)
+				return ctrl.Result{}, err
+			}
+			fmt.Printf("Got pfName %s matching PCI %s\n", pfName, *sriovConfig.PciAddr)
+			fmt.Printf("pfList before = %s\n", pfList)
+			pfList = append(pfList, pfName)
+			fmt.Printf("pfList after = %s\n", pfList)
+		} else if sriovConfig.VendorId != nil && sriovConfig.DeviceId != nil {
+			fmt.Printf("Configuring via device/vendor\n")
+			pfList, err = getPfListForVendorAndDevice(*sriovConfig.VendorId, *sriovConfig.DeviceId)
+			if err != nil {
+				fmt.Printf("Failed to find any devices with vendor %s device %s\n", *sriovConfig.VendorId, *sriovConfig.DeviceId)
+				return ctrl.Result{}, err
+			}
+		}
+		fmt.Printf("Got matching PFs: %v\n", pfList)
+		for _, pfName := range pfList {
+			if verifyPfExists(pfName) == false {
+				fmt.Printf("NIC %s does not exist on host, skipping...\n", pfName)
+				specApplied = false
+				continue
+			}
+			createVfsForPfName(pfName, *sriovConfig.NumVfs)
+
+			if sriovConfig.VfDriver != nil {
+				enableDriverForVfs(pfName, *sriovConfig.VfDriver)
+			} else {
+				// If driver field is omitted, set the default kernel driver
+				enableDriverForVfs(pfName, "ixgbevf")
+			}
+
+			if sriovConfig.MTU != nil && *sriovConfig.MTU > 0 {
+				if err := setMtuForPf(pfName, *sriovConfig.MTU); err != nil {
+					fmt.Printf("Failed setting MTU %d on %s\n", *sriovConfig.MTU, pfName)
+					return ctrl.Result{}, err
+				}
+			}
+		}
+	}
+	if specApplied == true {
+		fmt.Printf("Spec was applied on this host!\n")
+	}
+
+	hoststate.DiscoverHostState(r.NodeName, r.Client)
```

Comment:
Shouldn't this happen before you start applying the spec?

Reply:
It does, the main.go also invokes this first before statring the controller manager. Then we want to update the HostState after applying the spec.

In later phase, we'll make this a periodic task/ticker that monitors the host and periodically updates HostState.